ruby/phase-1.md

```diff
@@ -0,0 +1,536 @@
+    puts 'Hello Monkeys!'
+  end
+
+  def new_method
+    puts 'Hello New World!'
+  end
+end
+
+my = MyClass.new
+my.my_method
+# => Hello Monkeys!
+my.new_method
+# => Hello New World
+```
+
+### Concurrency
+
+Ruby can be a tricky language for running concurrent applications. It supports kernel threads,
+however it has a Global Interpreter Lock (GIL) in place which prevents any two threads to be
+executed at the same time (making it effectively single-threaded). The good news is that GIL does
+allow IO-bound instructions to be executed in parallel.
+
+While threads can offer a great performance boost in applications with a significant IO reliance. In
+order to effectively utilise a multi-core processor a Ruby application has to be executed using
+multiple processes.
+
+#### Fibers
+
+Ruby also has Fibers that allow light weight cooperative concurrency. These are code blocks that can
+be paused (from inside) and resumed (from outside). These are really useful for controlling the
+execution of workflow code. However unlike threads Fibers are never preempted and their scheduling
+has to be done manually by the code:
+
+```ruby
+fiber = Fiber.new do
+  i = 0
+  loop do
+    i += 1
+    puts i
+    Fiber.yield
+  end
+end
+
+fiber.resume
+# => 1
+fiber.resume
+# => 2
+fiber.resume
+# => 3
+```
+
+*NOTE: Each `fiber.resume` call blocks the main thread until a Fiber `yield`s back control.*
+
+#### Reactors
+
+Somewhat less popular, but still very viable are libraries that implement a reactor pattern for
+building concurrent applications (e.g. Async or EventMachine). These libraries allow users to
+concurrently execute IO operations.
+
+There are however quite a few big limitations to this approach:
+
+- You need to use very specific libraries to perform IO calls this way (there isn't one for gRPC)
+- Any CPU-bound instructions block the reactor and therefore need to be executed in small bursts per
+  reactor tick
+- Any user-defined code can block the whole thing and has to be wrapped in threads anyways
+
+
+## A basic end-to-end example
+
+Perhaps a bit unconventionally (as far as the proposal review process goes), but we'd like to start
+with an end-to-end example of a simple and fully functional "Hello World" workflow. The intention of
+this example is to demonstrate how everything ties together before diving into each separate moving
+part and exploring it in depth.
+
+```ruby
+require 'temporal'
+require 'temporal/worker'
+
+class MyActivity < Temporal::Activity
+  def execute(name)
+    puts "Hello World, #{name}!"
+  end
+end
+
+class MyWorkflow < Temporal::Workflow
+  def execute(name)
+    # This part is outside of the scope of the 1st phase of the proposal
+    MyActivity.execute!(name, options: { task_queue: 'my-queue' })
+  end
+end
+
+config = Temporal::Configuration.new(
+  url: 'localhost:7233'
+)
+
+client = Temporal::Client.new(config)
+client.start_workflow(MyWorkflow, 'Alice', options: {
+  namespace: 'my-namespace',
```

Comment:
I've mentioned this later in the proposal. There are a few reasons why I've decided not to limit the client:
1. You can implement a `NamespacedClient` based on the current `Client`, but not the other way round
2. Restricting Client will not allow me to implement execution attributes per workflow/activity class definition, which we decided to run as a proposal in phase 2

In case it's not a hard requirement let's go with it, but if there's strong pushback we'll find an alternative